ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx

```diff
@@ -18,128 +18,124 @@ interface WorkflowTimelineProps {
-        this.resizeSubscription = fromEvent(window, 'resize').subscribe(() => this.updateWidth());
-        this.updateWidth();
-    }
+    useEffect(() => {
+        resizeSubscription = fromEvent(window, 'resize').subscribe(updateWidth);
+        updateWidth();
 
-    public componentWillReceiveProps(nextProps: WorkflowTimelineProps) {
-        this.ensureRunningWorkflowRefreshing(nextProps.workflow);
-    }
+        return () => {
+            if (resizeSubscription) {
+                resizeSubscription.unsubscribe();
+            }
+            if (refreshSubscription) {
+                refreshSubscription.unsubscribe();
+            }
+        };
+    }, []);
 
-    public componentWillUnmount() {
-        if (this.resizeSubscription) {
-            this.resizeSubscription.unsubscribe();
-            this.resizeSubscription = null;
-        }
-        if (this.refreshSubscription) {
-            this.refreshSubscription.unsubscribe();
-            this.refreshSubscription = null;
+    useEffect(() => {
+        ensureRunningWorkflowRefreshing(props.workflow);
+    }, [props.workflow]);
+
+    function updateWidth() {
+        if (containerRef.current) {
+            setParentWidth((containerRef.current.offsetParent || window.document.body).clientWidth - NODE_NAME_WIDTH);
         }
     }
 
-    public render() {
-        if (!this.props.workflow.status.nodes) {
-            return <p>No nodes</p>;
-        }
-        const nodes = Object.keys(this.props.workflow.status.nodes)
-            .map(id => {
-                const node = this.props.workflow.status.nodes[id];
-                node.finishedAt = node.finishedAt || this.state.now.format();
-                node.startedAt = node.startedAt || this.state.now.format();
-                return node;
-            })
-            .filter(node => node.startedAt && node.type === 'Pod')
-            .sort((first, second) => {
-                const diff = moment(first.startedAt).diff(second.startedAt);
-                // If node started almost at the same time then sort by end time
-                if (diff <= 2) {
-                    return moment(first.finishedAt).diff(second.finishedAt);
-                }
-                return diff;
+    function ensureRunningWorkflowRefreshing(workflow: models.Workflow) {
+        const completedPhases = [models.NODE_PHASE.ERROR, models.NODE_PHASE.SUCCEEDED, models.NODE_PHASE.SKIPPED, models.NODE_PHASE.OMITTED, models.NODE_PHASE.FAILED];
+        const isCompleted = workflow && workflow.status && completedPhases.indexOf(workflow.status.phase) > -1;
+        if (!refreshSubscription && !isCompleted) {
+            refreshSubscription = interval(1000).subscribe(() => {
+                setNow(moment());
             });
-        if (nodes.length === 0) {
-            return <div />;
+        } else if (refreshSubscription && isCompleted) {
+            refreshSubscription.unsubscribe();
+            refreshSubscription = null;
```

Comment:
```suggestion
        if (!refreshSubscription.current && !isCompleted) {
            refreshSubscription.current = interval(1000).subscribe(() => {
                setNow(moment());
            });
        } else if (refreshSubscription.current && isCompleted) {
            refreshSubscription.current.unsubscribe();
            refreshSubscription.current = null;
```

will need to add `.current` here when it's a ref as well